table.cc

```diff
@@ -2381,3 +2392,55 @@ class table::table_state : public compaction::table_state {
 compaction::table_state& table::as_table_state() const noexcept {
     return *_table_state;
 }
+
+future<> table::update_repair_time(dht::token_range range, gc_clock::time_point repair_time) {
+    _repair_time_map->map += std::make_pair(locator::token_metadata::range_to_interval(std::move(range)), repair_time);
+    return make_ready_future<>();
+}
+
+gc_clock::time_point
+table::get_gc_before(const dht::decorated_key& dk, const gc_clock::time_point& query_time) const {
+    // if mode = timeout    // default option, if user does not specify tombstone_gc options
+    // if mode = disabled   // never gc tombstone
+    // if mode = immediate  // can gc tombstone immediately
+    // if mode = repair     // gc after repair
+    const auto& options = _schema->tombstone_gc_options();
+    switch (options.mode()) {
+    case tombstone_gc_mode::timeout:
+        tlogger.info("Get gc_before for ks={}, table={}, dk={}, mode=timeout", schema()->ks_name(), schema()->cf_name(), dk);
+        return saturating_subtract(query_time, _schema->gc_grace_seconds());
+    case tombstone_gc_mode::disabled:
+        tlogger.info("Get gc_before for ks={}, table={}, dk={}, mode=disabled", schema()->ks_name(), schema()->cf_name(), dk);
+        return gc_clock::time_point::min();
+    case tombstone_gc_mode::immediate:
+        tlogger.info("Get gc_before for ks={}, table={}, dk={}, mode=immediate", schema()->ks_name(), schema()->cf_name(), dk);
+        return gc_clock::time_point::max();
+    case tombstone_gc_mode::repair:
+        const std::chrono::seconds& propagation_delay = options.propagation_delay_in_seconds();
+        auto gc_before = gc_clock::time_point::min();
+        auto repair_timestamp = gc_clock::time_point::min();
+        if (!needs_repair_before_gc()) {
+            repair_timestamp = gc_clock::now();
+            gc_before = saturating_subtract(repair_timestamp, propagation_delay);
+        } else {
+            const auto it = _repair_time_map->map.find(dk.token());
+            if (it == _repair_time_map->map.end()) {
+                gc_before = gc_clock::time_point::min();
+            } else {
+                repair_timestamp = it->second;
+                gc_before = saturating_subtract(repair_timestamp, propagation_delay);
+            }
+        }
+        tlogger.info("Get gc_before for ks={}, table={}, dk={}, mode=repair, repair_timestamp={}, propagation_delay={}, gc_before={}",
+                schema()->ks_name(), schema()->cf_name(), dk, repair_timestamp, propagation_delay.count(), gc_before);
+        return gc_before;
+    }
+}
```

Comment:
How often is it called? I'd like to understand the impact on compaction speed.

If we return struct { interval<dht::token> for_tokens, gc_clock::time_point gc_before } we only have to call it when the interval changes.

Reply:
Avi, do not worry. We have discussed this and agreed to do as follow ups. There is a nice opportunity to optimize to reduce the search because compaction are sequential.